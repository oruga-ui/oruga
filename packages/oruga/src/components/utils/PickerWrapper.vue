<script setup lang="ts">
import { computed, useAttrs, type PropType, ref, watch, nextTick } from "vue";

import ODropdown from "../dropdown/Dropdown.vue";
import ODropdownItem from "../dropdown/DropdownItem.vue";
import OInput from "../input/Input.vue";

import { isMobileAgent } from "@/utils/helpers";
import { isClient } from "@/utils/ssr";
import {
    getActiveClasses,
    useEventListener,
    useInputHandler,
    usePropBinding,
} from "@/composables";

import type { ClassBind, ComponentClass } from "@/types";

/**
 * This is a internal used component.
 * Used by Datepicker and Timepicker.
 */
defineOptions({
    name: "OPickerWrapper",
});

const props = defineProps({
    /** parent picker component props  */
    pickerProps: { type: Object, required: true },
    /** data-oruga attribute value */
    dataOruga: { type: String, required: true },
    /** the input value */
    value: { type: [Date, Array], default: undefined },
    /** the active state of the dropdown */
    active: { type: Boolean, default: false },
    formattedValue: { type: String, default: undefined },
    nativeType: { type: String, required: true },
    nativeStep: { type: String, default: undefined },
    nativeValue: { type: [String, Number], default: undefined },
    nativeMin: { type: [String, Number], default: undefined },
    nativeMax: { type: [String, Number], default: undefined },
    stayOpen: { type: Boolean, default: false },
    rootClasses: {
        type: Array as PropType<ClassBind[]>,
        required: true,
    },
    dropdownClasses: {
        type: Array as PropType<ClassBind[]>,
        required: true,
    },
    boxClass: {
        type: Array as PropType<ComponentClass>,
        required: true,
    },
});

const emits = defineEmits<{
    /**
     * active prop two-way binding
     * @param value {boolean} updated active prop
     */
    (e: "update:active", value: boolean): void;
    /** on value change event */
    (e: "change", value: string): void;
    /** on natvie value change event */
    (e: "native-change", value: string): void;
    /** on input focus event */
    (e: "focus", evt: Event): void;
    /** on input blur event */
    (e: "blur", evt: Event): void;
    /** on input invalid event */
    (e: "invalid", evt: Event): void;
    /** on icon click event */
    (e: "icon-click", evt: Event): void;
    /** on icon right click event */
    (e: "icon-right-click", evt: Event): void;
}>();

/** the computed picker contains all chared props from the datepicker and the timepicker  */
const picker = computed<any>(() => props.pickerProps);

const isMobileNative = computed(
    () => picker.value.mobileNative && isMobileAgent.any(),
);

const dropdownRef = ref<InstanceType<typeof ODropdown>>();
const inputRef = ref<InstanceType<typeof OInput>>();
const nativeInputRef = ref<InstanceType<typeof OInput>>();

const elementRef = computed(() =>
    isMobileNative.value ? nativeInputRef.value : inputRef.value,
);

// use form input functionality for native input
const {
    checkHtml5Validity,
    setFocus,
    doClick,
    onBlur,
    onFocus,
    onInvalid,
    isValid,
    isFocused,
} = useInputHandler(elementRef, emits, picker.value);

/**
 * Show input as text for placeholder,
 * when placeholder and native value is given and input is not focused.
 */
const computedNativeType = computed(() =>
    !picker.value.placeholder || props.nativeValue || isFocused.value
        ? props.nativeType
        : "text",
);

/**
 * When v-model is changed:
 *  1. Update internal value.
 *  2. If it's invalid, validate again.
 */
watch(
    () => props.value,
    () => {
        // toggle picker if not stay open
        if (!props.stayOpen) togglePicker(false);
        // check validation if native
        if (isMobileNative.value && !isValid.value) checkHtml5Validity();
    },
);

const isActive = usePropBinding<boolean>("active", props, emits, {
    passive: true,
});

watch(isActive, onActiveChange);

const ariaRole = computed(() => (!picker.value.inline ? "dialog" : undefined));

const triggers = computed(() => (picker.value.openOnFocus ? ["click"] : []));

if (isClient) useEventListener("keyup", onKeyPress);

/** Keypress event that is bound to the document. */
function onKeyPress(event: KeyboardEvent): void {
    if (isActive.value && (event.key === "Escape" || event.key === "Esc"))
        togglePicker(false);
}

// --- EVENT HANDLER ---

/** Toggle picker */
function togglePicker(active: boolean): void {
    if (isMobileNative.value) {
        setFocus(); // focus the underlaying input element
        doClick(); // click to open the underlaying input element
    } else if (dropdownRef.value) {
        if (active || picker.value.closeOnClick)
            nextTick(() => (isActive.value = active));
    }
}

/** Avoid dropdown toggle when is already visible */
function onInputClick(event): void {
    if (isActive.value) event.stopPropagation();
}

/** Emit 'blur' event on dropdown is not active (closed) */
function onActiveChange(value: boolean): void {
    if (value) onFocus();
    else if (!value) onBlur();
}

// --- Computed Component Classes ---

const attrs = useAttrs();
const inputBind = computed(() => ({
    ...attrs,
    ...picker.value.inputClasses,
}));

const dropdownBind = computed(() => ({
    "root-class": getActiveClasses(props.dropdownClasses),
    ...picker.value.dropdownClasses,
}));

// --- Expose Public Functionalities ---

/** expose functionalities for programmatic usage */
defineExpose({ focus: setFocus });
</script>

<template>
    <div :data-oruga="dataOruga" :class="rootClasses">
        <o-dropdown
            v-if="!isMobileNative || picker.inline"
            ref="dropdownRef"
            v-bind="dropdownBind"
            v-model:active="isActive"
            :position="picker.position"
            :disabled="picker.disabled"
            :inline="picker.inline"
            :mobile-modal="picker.mobileModal"
            :trap-focus="picker.trapFocus"
            :aria-role="ariaRole"
            :aria-modal="!picker.inline"
            :tabindex="-1"
            :teleport="picker.teleport"
            :triggers="triggers">
            <template v-if="!picker.inline" #trigger>
                <slot name="trigger">
                    <o-input
                        ref="inputRef"
                        v-bind="inputBind"
                        autocomplete="off"
                        :model-value="formattedValue"
                        :placeholder="picker.placeholder"
                        :size="picker.size"
                        :icon-pack="picker.iconPack"
                        :icon="picker.icon"
                        :icon-right="picker.iconRight"
                        :icon-right-clickable="picker.iconRightClickable"
                        :expanded="picker.expanded"
                        :rounded="picker.rounded"
                        :disabled="picker.disabled"
                        :readonly="picker.readonly"
                        :use-html5-validation="false"
                        @click="onInputClick"
                        @keyup.enter="togglePicker(true)"
                        @change="$emit('change', $event.target.value)"
                        @focus="onFocus"
                        @blur="onBlur"
                        @icon-click="$emit('icon-click', $event)"
                        @icon-right-click="$emit('icon-right-click', $event)" />
                </slot>
            </template>
            <o-dropdown-item
                override
                tag="div"
                :item-class="boxClass"
                :disabled="picker.disabled"
                :clickable="false">
                <slot />
            </o-dropdown-item>
        </o-dropdown>

        <!-- Native Picker -->
        <o-input
            v-else
            ref="nativeInputRef"
            v-bind="inputBind"
            :type="computedNativeType"
            autocomplete="off"
            :model-value="nativeValue"
            :min="nativeMin"
            :max="nativeMax"
            :step="nativeStep"
            :placeholder="picker.placeholder"
            :size="picker.size"
            :icon-pack="picker.iconPack"
            :icon="picker.icon"
            :icon-right="picker.iconRight"
            :icon-right-clickable="picker.iconRightClickable"
            :rounded="picker.rounded"
            :disabled="picker.disabled"
            :readonly="false"
            :use-html5-validation="false"
            @change="$emit('native-change', $event.target.value)"
            @focus="onFocus"
            @blur="onBlur"
            @invalid="onInvalid"
            @icon-click="$emit('icon-click', $event)"
            @icon-right-click="$emit('icon-right-click', $event)" />
    </div>
</template>
