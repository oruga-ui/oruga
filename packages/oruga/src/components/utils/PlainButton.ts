import { createVNode, type FunctionalComponent } from "vue";

/**
 * Acts much like a regular `<button>` element except that its type defaults to "button" even inside forms
 *
 * Useful as a default tag for components that use custom action buttons
 */
const PlainButton: FunctionalComponent = (props, { slots }) =>
    createVNode(
        "button",
        {
            type: "button",
            ...props,
        },
        slots.default?.(),
    );
export default PlainButton;
