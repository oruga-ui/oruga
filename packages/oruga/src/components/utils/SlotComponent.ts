import type { DefineComponent, PropType } from "vue";
import { defineComponent, h } from "vue";
import type { DynamicComponent } from "@/types";

/** This components renders a specific slot and only the slot of another component */
export default defineComponent({
    name: "OSlotComponent",
    props: {
        /** Component to be get the slot from */
        component: { type: Object, required: true },
        /** Slot name */
        name: { type: String, default: "default" },
        /** Props passed to the slot */
        props: { type: Object, default: () => {} },
        /** Tag name of the slot wrapper element */
        tag: {
            type: [String, Object, Function] as PropType<DynamicComponent>,
            default: "div" as DynamicComponent,
        },
    },
    render() {
        const slot = (this.component as DefineComponent).$slots[this.name]
            ? (this.component as DefineComponent).$slots[this.name](this.props)
            : {};

        return h(this.tag, {}, slot);
    },
});
