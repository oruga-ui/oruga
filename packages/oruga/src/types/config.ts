import type { ValidatableFormElement } from "@/composables";
import type { IconConfig } from "@/utils/icons";
import type { DynamicComponent } from "./utils";

export type ComponentProps = Record<string, any>;

export type ClassFunction = (
    suffix: string,
    props: ComponentProps,
) => string | undefined;

export type ClassObject = {
    class: string;
    /**
     * In case you want to override existing Oruga default classes for this class,
     * you can set the field override to true.
     */
    override?: boolean;
};

export type TransformFunction = (classes: string) => string;

export type ComponentClass = string | Array<string> | ClassFunction;

export type ClassDefinition =
    | string
    | Array<string>
    | ClassFunction
    | ClassObject;

export type ComponentConfigBase = Partial<{
    /**
     * In case you want to override existing Oruga default classes for this component,
     * you can set the field override to true.
     */
    override: boolean;
    /** Use this function in case you want to transform all applied classes' names. */
    transformClasses: TransformFunction;
}>;

export type GlobalConfig = Partial<{
    /** Define the icon pack be used */
    iconPack: string;
    /** Define custom icon packs */
    customIconPacks: Record<string, IconConfig>;
    /** Define a specific icon component */
    iconComponent: DynamicComponent;
    /** Enable HTML5 validation */
    useHtml5Validation: boolean;
    /** Show input status icon using field and variant prop */
    statusIcon: boolean;
    /** Custom HTML5 validation invalid handler */
    invalidHandler: (
        validatable: ValidatableFormElement,
        fieldElement: Element,
    ) => void;
    /** You can use transformClasses globally if you need to transform classes for any component. */
    transformClasses: TransformFunction;
    /**
     * In case you want to override existing Oruga default classes all components completely,
     * you can specify the override behaviour here globaly.
     */
    override: boolean;
}>;

export interface OrugaOptions extends GlobalConfig {}
