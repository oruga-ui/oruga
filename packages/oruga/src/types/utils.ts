import type { Component, EmitsToProps } from "vue";
import type { ComponentEmit } from "vue-component-type-helpers";

export type ClassBind = {
    [x: string]: boolean;
};

/** Vue native type which a v-bind prop must have */
export type PropBind =
    | string
    | {
          [x: string]: string;
      }
    | {
          [x: string]: boolean;
      };

/** Vue native dynamic component 'is' property value type */
export type DynamicComponent = string | object | CallableFunction | Component;

/** Define a property as required */
export type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] };

/** Custom type helper which extracts the `$emits` type of an component and converts it to an props object. */
export type ComponentEmits<C> = EmitsToProps<ComponentEmit<C>>;

/** Extracts the type of property `K` in object `T`. */
type TypeOfKey<T, K extends string> = K extends keyof T ? T[K] : unknown;

/**
 * Extracts the type of a deep property `K` within an object `T`.
 * Property `K` can be defined as a property path with `.` seperator.
 */
export type DeepType<T, K> = T extends object
    ? K extends string
        ? K extends `${infer F}.${infer R}`
            ? DeepType<TypeOfKey<T, F>, R>
            : TypeOfKey<T, K>
        : T
    : T;
