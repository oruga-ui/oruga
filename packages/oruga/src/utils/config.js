import { merge } from "./helpers"

let config = {
    iconPack: 'mdi',
    useHtml5Validation: true,
    statusIcon: true
}

export const setOptions = (options) => { config = options }

export const getOptions = () => { return config }

export let VueInstance

export const setVueInstance = (Vue) => { VueInstance = Vue }

export const Programmatic = {
    getOptions,
    setOptions(options) {
        setOptions(merge(getOptions(), options, true))
    }
}

export const Plugin = {
    install(Vue, options = {}) {
        setVueInstance(Vue)
        // Options
        setOptions(merge(getOptions(), options, true))
    }
}
