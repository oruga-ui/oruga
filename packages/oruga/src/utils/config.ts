import { ref, toRaw, type App } from "vue";
import { getValueByPath, merge, setValueByPath } from "./helpers";
import { setVueInstance } from "./plugins";
import type { DeepType, OrugaOptions } from "@/types";

declare module "../index" {
    interface OrugaProgrammatic {
        config: typeof ConfigProgrammatic;
    }
}

const globalOptions = ref<OrugaOptions>({
    iconPack: "mdi",
    useHtml5Validation: true,
    statusIcon: true,
    transformClasses: undefined,
});

export const setOptions = (options: OrugaOptions): void => {
    globalOptions.value = options;
};

export const getOptions = (): OrugaOptions => {
    return Object.assign({}, toRaw(globalOptions.value));
};

export const getOption = <K extends string = string>(
    path: K,
    defaultValue?: DeepType<OrugaOptions, K>,
): DeepType<OrugaOptions, K> => {
    return getValueByPath(globalOptions.value, path, defaultValue);
};

/** less type strict version of getOption for component props defaults */
export const getDefault = <T>(path: string, defaultValue?: T): T => {
    return getValueByPath(globalOptions.value, path, defaultValue) as T & {};
};

export const setOption = <T>(path: string, value: T): void => {
    setValueByPath(globalOptions.value, path, value);
};

export const ConfigProgrammatic = {
    getOption,
    getOptions,
    setOption,
    setOptions(options: OrugaOptions): void {
        setOptions(merge(getOptions(), options, true));
    },
};

export const OrugaConfig = {
    install(app: App, options?: OrugaOptions): void {
        // set global vue instance
        setVueInstance(app);
        // set options
        setOptions(merge(getOptions(), options, true));
    },
};
