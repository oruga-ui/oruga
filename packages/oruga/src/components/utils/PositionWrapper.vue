<script setup lang="ts">
import {
    watch,
    computed,
    nextTick,
    onBeforeUnmount,
    ref,
    type Component,
    type PropType,
} from "vue";
import { isClient } from "@/utils/ssr";
import { getScrollingParent, unrefElement } from "@/composables";

type Position = "top" | "bottom" | "left" | "right";

const opposites: Record<Position, Position> = {
    top: "bottom",
    bottom: "top",
    right: "left",
    left: "right",
};

type Point = { x: number; y: number };

defineOptions({ inheritAttrs: false });

const props = defineProps({
    teleport: {
        type: [Boolean, String, Object],
        required: true,
    },
    trigger: {
        type: Object as PropType<HTMLElement | Component>,
        default: undefined,
    },
    content: {
        type: Object as PropType<HTMLElement | Component>,
        default: undefined,
    },
    /**
     * Position of the component relative to the trigger
     * @values auto, top, bottom, left, right, top-right, top-left, bottom-left, bottom-right
     */
    position: {
        type: String,
        validator: (value: string) =>
            [
                "auto",
                "top",
                "bottom",
                "left",
                "right",
                "top-right",
                "top-left",
                "bottom-left",
                "bottom-right",
            ].indexOf(value) > -1,
        default: undefined,
    },
    /** Used for calculation position auto */
    defaultPosition: {
        type: String as PropType<Position>,
        validator: (value: string) =>
            ["top", "bottom", "left", "right"].indexOf(value) > -1,
        default: "top",
    },
    /** disable the position calculation */
    disabled: { type: Boolean, default: false },
    /** update positioning on teleport */
    disablePositioning: { type: Boolean, default: true },
});

const emits = defineEmits<{
    (e: "update:position", value: string);
}>();

const to = computed(() =>
    typeof props.teleport === "boolean" ? "body" : props.teleport,
);

const disabled = computed(() =>
    typeof props.teleport === "boolean" || !props.teleport
        ? !props.teleport
        : false,
);

// --- Dynamic Positioning Handling Feature ---

const initialPosition = props.position;

const scrollingParent = ref(undefined);
const resizeObserver = ref(null);

if (isClient && window.ResizeObserver) {
    resizeObserver.value = new window.ResizeObserver(updatePositioning);
}

// on content or disable state change update event listener
watch(
    [() => props.disabled, () => props.content],
    () => {
        if (!props.disabled) addHandler();
        else removeHandler();
    },
    { immediate: true },
);

// update positioning if props change
watch(
    [
        () => props.trigger,
        () => props.content,
        () => props.disablePositioning,
        () => props.disabled,
    ],
    () => {
        nextTick(() => updatePositioning());
    },
    { immediate: true },
);

// remove any event listener on unmount
onBeforeUnmount(() => removeHandler());

/** add event listener */
function addHandler(): void {
    if (isClient && !scrollingParent.value && props.content) {
        // get parent container
        scrollingParent.value = getScrollingParent(unrefElement(props.content));
        // set event listener
        if (
            scrollingParent.value &&
            scrollingParent.value !== document.documentElement
        ) {
            scrollingParent.value.addEventListener(
                "scroll",
                updatePositioning,
                { passive: true },
            );
            if (window.ResizeObserver)
                resizeObserver.value.observe(scrollingParent.value);
        } else {
            document.addEventListener("scroll", updatePositioning, {
                passive: true,
            });
            window.addEventListener("resize", updatePositioning);
        }
    }
}

/** remove event listener */
function removeHandler(): void {
    if (isClient) {
        if (window.ResizeObserver) resizeObserver.value?.disconnect();
        window.removeEventListener("resize", updatePositioning);
        document.removeEventListener("scroll", updatePositioning);
        scrollingParent.value = undefined;
    }
}

/** Update the best position set teleport positioning */
function updatePositioning(): void {
    if (props.disabled) return;
    let position = props.position;
    // update position if auto position is enabled
    if (initialPosition === "auto") {
        // calculate best position
        position = getAutoPosition();
        if (position != props.position)
            // two-way bind updated position
            emits("update:position", position);
    }
    // do not set content position if not teleport enabled
    if (!props.teleport) return;

    const content = unrefElement(props.content);
    const trigger = unrefElement(props.trigger);

    // set content position
    if (content && trigger) {
        const rect = trigger.getBoundingClientRect();
        let top = rect.top + window.scrollY;
        let left = rect.left + window.scrollX;
        // define vertical positioning
        if (position.includes("bottom")) {
            top += trigger.clientHeight;
        } else if (position.includes("top")) {
            top -= content.clientHeight;
        }
        // define horizontal positioning
        if (position === "left") {
            left -= content.clientWidth;
        } else if (position === "right") {
            left += trigger.clientWidth;
        } else if (position.includes("-right")) {
            left += trigger.clientWidth - content.clientWidth;
        }

        // adjust exact vertical positioning
        if (position === "top" || position === "bottom") {
            left += trigger.clientWidth / 2; //- content.clientWidth / 2;
        }
        // adjust exact horizontal positioning
        if (position === "left" || position === "right") {
            top += trigger.clientHeight / 2; // - content.clientHeight / 2;
        }

        // set style properties
        if (props.disablePositioning) {
            content.style.position = "relative";
            content.style.top = `${top}px`;
            content.style.left = `${left}px`;
        } else {
            content.style.position = "";
            content.style.top = "";
            content.style.left = "";
        }
    }
}

/** calculate best position if auto */
function getAutoPosition(): string {
    let bestPosition = props.defaultPosition;
    if (!props.content || !props.trigger) return bestPosition;
    if (!scrollingParent.value) return bestPosition;

    // get viewport from container
    const viewRect = new DOMRect(
        scrollingParent.value.offsetLeft,
        scrollingParent.value.offsetTop,
        scrollingParent.value.clientWidth,
        scrollingParent.value.clientHeight,
    );

    const contentRect = unrefElement(props.content).getBoundingClientRect();
    const triggerRect = unrefElement(props.trigger).getBoundingClientRect();

    // detect auto position
    const triggerAnchors = anchors(triggerRect);
    const contentAnchors = anchors(contentRect);
    const contentRectAtAnchor = (pos: Position) => {
        const triggerAnchor = triggerAnchors[pos];
        const contentAnchor = contentAnchors[opposites[pos]];
        // Translates contentRect so contentAnchor is on top of triggerAnchor
        // NOTE: this doesn't account for the extra offset that the tooltip arrow provides.
        // That offset should be small, and it's tricky to get it from the CSS.
        return new DOMRect(
            contentRect.x + (triggerAnchor.x - contentAnchor.x),
            contentRect.y + (triggerAnchor.y - contentAnchor.y),
            contentRect.width,
            contentRect.height,
        );
    };

    const defaultOpposite = opposites[props.defaultPosition];
    const crossPosition =
        props.defaultPosition === "top" || props.defaultPosition === "bottom"
            ? "left"
            : "top";
    const crossOpposite = opposites[crossPosition];
    // In descending order of priority
    const positions: Position[] = [
        props.defaultPosition,
        defaultOpposite,
        crossPosition,
        crossOpposite,
    ];
    let maxOverlap = 0;
    for (const position of positions) {
        const overlap = intersectionArea(
            viewRect,
            contentRectAtAnchor(position),
        );
        if (overlap > maxOverlap) {
            maxOverlap = overlap;
            bestPosition = position;
        }
    }
    return bestPosition;
}

// --- Helper Functions ---

function intersectionArea(a: DOMRect, b: DOMRect): number {
    const left = Math.max(a.left, b.left);
    const right = Math.min(a.right, b.right);
    const top = Math.max(a.top, b.top);
    const bottom = Math.min(a.bottom, b.bottom);
    return Math.max(right - left, 0) * Math.max(bottom - top, 0);
}

/**
 * @param rect the bounding rectangle of the trigger element
 * @return the "anchor points" (points where the arrow attaches) for each side of the tooltip
 */
const anchors = (rect: DOMRect): Record<Position, Point> => ({
    top: { x: (rect.left + rect.right) * 0.5, y: rect.top },
    bottom: { x: (rect.left + rect.right) * 0.5, y: rect.bottom },
    left: { x: rect.left, y: (rect.top + rect.bottom) * 0.5 },
    right: { x: rect.right, y: (rect.top + rect.bottom) * 0.5 },
});
</script>

<template>
    <Teleport :to="to" :disabled="disabled">
        <template v-if="disabled">
            <slot />
        </template>
        <template v-else>
            <div
                v-bind="$attrs"
                :style="{ position: 'absolute', left: '0px', top: '0px' }">
                <slot />
            </div>
        </template>
    </Teleport>
</template>
