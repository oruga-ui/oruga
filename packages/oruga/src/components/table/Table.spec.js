import Vue from 'vue'
import { shallowMount } from '@vue/test-utils'
import { useFakeTimers } from 'sinon'
import OInput from '@components/input/Input'
import OTable from '@components/table/Table'
import { setVueInstance } from '../../utils/config'

describe('OTable', () => {
    setVueInstance(Vue)

    let wrapper
    beforeEach(() => {
        wrapper = shallowMount(OTable)
    })

    const tableCols = shallowMount(OTable, {
        propsData: {
            columns: [
                { label: 'default', width: '100px' },
                { label: 'pecent', width: '50%' },
                { label: 'fixed_num', width: 100 },
                { label: 'fixed_str', width: '100' }
            ]
        }
    })

    it('is called', () => {
        expect(wrapper.exists()).toBeTruthy()
    })

    it('has the filter row visible when searchable', () => {
        wrapper.setProps({
            columns: [
                {
                    field: 'id',
                    label: 'ID',
                    width: '40',
                    numeric: true
                }
            ]
        })
        // Don't show if no searchable column
        expect(wrapper.vm.hasSearchablenewColumns).toBe(false)
        // Show if one or more searchable column
        wrapper.setProps({
            columns: [
                {
                    field: 'id',
                    label: 'ID',
                    width: '40',
                    numeric: true,
                    searchable: true
                }
            ]
        })
        expect(wrapper.vm.hasSearchablenewColumns).toBe(true)
    })

    it('render correctly', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('holds columns', () => {
        const headers = tableCols.findAll('th')

        expect(headers.length).toBeGreaterThanOrEqual(4)

        const cols = headers.filter((th) => {
            return th.find('span')
        })

        expect(cols.length).toBe(4)
        expect(cols.at(0).attributes('style')).toBe('width: 100px;')
        expect(cols.at(1).attributes('style')).toBe('width: 50%;')
        expect(cols.at(2).attributes('style')).toBe('width: 100px;')
        expect(cols.at(3).attributes('style')).toBe('width: 100px;')
    })

    describe('Searchable', () => {
        const data = [
            { id: 1, name: 'Jesse' },
            { id: 2, name: 'João' },
            { id: 3, name: 'Tina' },
            { id: 4, name: 'Anne' },
            { id: 5, name: 'Clarence' }
        ]
        let headRows
        let bodyRows
        let searchInput

        beforeEach(() => {
            wrapper = shallowMount(OTable, {
                propsData: {
                    columns: [
                        { label: 'ID', field: 'id', numeric: true },
                        { label: 'Name', field: 'name', searchable: true }
                    ],
                    data
                }
            })
            headRows = wrapper.findAll('thead tr')
            bodyRows = wrapper.findAll('tbody tr')
            searchInput = wrapper.find(OInput)
        })

        it('displays filter row when at least one column is searchable', () => {
            expect(headRows).toHaveLength(2)
        })

        it('displays filter input only on searchable columns', () => {
            const filterCells = headRows.at(1).findAll('th')

            expect(filterCells.at(0).isEmpty()).toBe(true) // ID column is not searchable
            expect(filterCells.at(1).contains(OInput)).toBe(true) // Name column is searchable
        })

        it('displays all data', () => {
            expect(bodyRows).toHaveLength(5)
        })

        it('displays filtered data when searching', async () => {
            searchInput.vm.$emit('input', 'J')
            await wrapper.vm.$nextTick()

            bodyRows = wrapper.findAll('tbody tr')
            expect(bodyRows).toHaveLength(2) // Jesse and João
        })

        it('displays filtered data when searching and updating data', async () => {
            searchInput.vm.$emit('input', 'J')
            wrapper.setProps({
                data: [
                    ...data,
                    { id: 6, name: 'Justin' }
                ]
            })
            await wrapper.vm.$nextTick()

            bodyRows = wrapper.findAll('tbody tr')
            expect(bodyRows).toHaveLength(3) // Jesse, João and Justin
        })

        it('displays filtered data when searching by name without accent', async () => {
            searchInput.vm.$emit('input', 'Joao')
            await wrapper.vm.$nextTick()
            bodyRows = wrapper.findAll('tbody tr')
            expect(bodyRows).toHaveLength(1) // João
        })

        it('displays filtered data when searching by name with accent', async () => {
            searchInput.vm.$emit('input', 'João')
            await wrapper.vm.$nextTick()
            bodyRows = wrapper.findAll('tbody tr')
            expect(bodyRows).toHaveLength(1) // João
        })

        it('debounce search filtering when debounce-search is defined', () => {
            const clock = useFakeTimers()
            wrapper.setProps({
                debounceSearch: 1000
            })
            for (let i = 0; i < 10; i++) {
                searchInput.vm.$emit('input', 'J'.repeat(10 - i))
                clock.tick(500)
                bodyRows = wrapper.findAll('tbody tr')
                expect(bodyRows).toHaveLength(5) // No filtering yet
            }
            clock.tick(1000)
            bodyRows = wrapper.findAll('tbody tr')
            expect(bodyRows).toHaveLength(5) // Filtering after debounce
        })

        it('tests checkRow method',()=>{
            const shiftCheckRow = jest.spyOn(wrapper.vm,'shiftCheckRow')
            const isRowChecked = jest.spyOn(wrapper.vm,'isRowChecked')
            const removeCheckedRow = jest.spyOn(wrapper.vm,'removeCheckedRow')
            const checkRow = jest.spyOn(wrapper.vm,'checkRow')
            shiftCheckRow.mockImplementationOnce(()=>{})
            isRowChecked.mockImplementationOnce(()=>false)
            removeCheckedRow.mockImplementationOnce(()=>{})
            const index = 0
            let event = {
                shiftKey:true
            }
            wrapper.vm.lastCheckedRowIndex = 1
            const row = {
                id: 1,
                first_name: 'Jesse',
                last_name: 'Simmons',
                date: '2016-10-15 13:43:27',
                gender: 'Male'
              }
            wrapper.vm.checkRow(row,index,event)
            expect(checkRow).toHaveBeenCalledWith(row, 0, event)
            expect(wrapper.emitted().check[0]).toEqual([[],row])
            expect(wrapper.emitted()["update:checkedRows"][0]).toEqual([[]])
            event = {
                shiftKey:false
            }
            wrapper.vm.checkRow(row,index,event)
            expect(checkRow).toHaveBeenCalledWith(row, 0, event)
            expect(wrapper.emitted().check[0]).toEqual([[row],row])
            expect(wrapper.emitted()["update:checkedRows"][0]).toEqual([[row]])
            isRowChecked.mockImplementationOnce(()=>true)
            wrapper.vm.checkRow(row,index,event)
            expect(removeCheckedRow).toHaveBeenCalledWith(row)
        })

        it('tests isAllUncheckable method',async()=>{
            await wrapper.setProps({
                paginated:false,
                data:[{
                    id: 1,
                    first_name: 'Jesse',
                    last_name: 'Simmons',
                    date: '2016-10-15 13:43:27',
                    gender: 'Male'
                  }]
            })
            const isRowCheckable = jest.spyOn(wrapper.vm,'isRowCheckable')
            isRowCheckable.mockImplementationOnce(()=>true)
            expect(wrapper.vm.isAllUncheckable).toEqual(false)
        })

        it('tests total watcher ',async()=>{
            await wrapper.setProps({
                backendPagination:true,
                total:100
            })
            expect(wrapper.vm.newDataTotal).toEqual(100)
        })

        it('tests currentPage watcher ',async()=>{
            await wrapper.setProps({
                currentPage:5
            })
            expect(wrapper.vm.newCurrentPage).toEqual(5)
        })
    })
})
