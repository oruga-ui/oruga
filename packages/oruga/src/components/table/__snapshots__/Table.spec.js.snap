// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OTable render correctly 1`] = `
<div class="o-table__root">
    <div style="display: none;"></div>
    <!---->
    <!---->
    <div class="o-table__wrapper">
        <table class="o-table o-table--table__empty">
            <!---->
            <!---->
            <tbody>
                <tr>
                    <td colspan="0"></td>
                </tr>
            </tbody>
            <!---->
        </table>
        <o-loading-stub animation="fade" oncancel="() => {
          /* istanbul ignore next */
          cov_2jtz8vshvv.f[3]++;
        }" icon="loading" iconspin="true" iconsize="medium" overlay="true"></o-loading-stub>
    </div>
    <!---->
</div>
`;
